dask_planner/src/sql/optimizer/join_reorder.rs

```diff
@@ -0,0 +1,392 @@
+//! Join reordering based on the paper "Improving Join Reordering for Large Scale Distributed Computing"
+//! https://ieeexplore.ieee.org/document/9378281
+
+use std::collections::HashSet;
+
+use datafusion_common::{Column, Result};
+use datafusion_expr::{Expr, Join, JoinType, LogicalPlan, LogicalPlanBuilder};
+use datafusion_optimizer::{utils, utils::split_conjunction, OptimizerConfig, OptimizerRule};
+use log::warn;
+
+use crate::sql::table::DaskTableSource;
+
+pub struct JoinReorder {
+    /// Maximum number of fact tables to allow in a join
+    max_fact_tables: usize,
+    /// Ratio of the size of the dimension tables to fact tables
+    fact_dimension_ratio: f64,
+    /// Whether to preserve user-defined order of unfiltered dimensions
+    preserve_user_order: bool,
+    /// Constant to use when determining the number of rows produced by a
+    /// filtered relation
+    filter_selectivity: f64,
+}
+
+impl Default for JoinReorder {
+    fn default() -> Self {
+        Self {
+            max_fact_tables: 2,
+            // FIXME: fact_dimension_ratio should be 0.3
+            fact_dimension_ratio: 0.7,
+            preserve_user_order: true,
+            filter_selectivity: 1.0,
+        }
+    }
+}
+
+impl OptimizerRule for JoinReorder {
+    fn name(&self) -> &str {
+        "join_reorder"
+    }
+
+    fn try_optimize(
+        &self,
+        plan: &LogicalPlan,
+        _config: &dyn OptimizerConfig,
+    ) -> Result<Option<LogicalPlan>> {
+        let original_plan = plan.clone();
+        // Recurse down first
+        // We want the equivalent of Spark's transformUp here
+        let plan = utils::optimize_children(self, plan, _config)?;
+
+        match &plan {
+            Some(LogicalPlan::Join(join)) if join.join_type == JoinType::Inner => {
+                optimize_join(self, plan.as_ref().unwrap(), join)
+            }
+            Some(plan) => Ok(Some(plan.clone())),
+            None => match &original_plan {
+                LogicalPlan::Join(join) if join.join_type == JoinType::Inner => {
+                    optimize_join(self, &original_plan, join)
+                }
+                _ => Ok(None),
+            },
+        }
+    }
+}
+
+fn optimize_join(
+    rule: &JoinReorder,
+    plan: &LogicalPlan,
+    join: &Join,
+) -> Result<Option<LogicalPlan>> {
+    // FIXME: Check fact/fact join logic
+
+    if !is_supported_join(join) {
+        return Ok(Some(plan.clone()));
+    }
+
+    // Extract the relations and join conditions
+    let (rels, conds) = extract_inner_joins(plan);
+
+    // Split rels into facts and dims
+    let rels: Vec<Relation> = rels.into_iter().map(Relation::new).collect();
```

Comment:
Nope, could totally be done in the `extract_inner_joins` function. Do you have a preference here?

Reply:
My preference would be to do this in `extract_inner_joins` (if doing so isn't particularly nasty), since we seemingly don't have any use for the returned `Vec<LogicalPlan>` past this point